### PR TITLE
Proved inverted inputs and pullup mode

### PR DESCRIPTION
**Pull request by Beerlesklopfer**
Hi Bertrand,

this is the output of the "InvertedInputs.ino" sketch. So it is proved to work correctly.

Best gregards Joerg

##### Checking setting port mode #####

------------- SUCCEEDED --------------

#####   INTERRUPT PORT B FIRED   #####

------------- SUCCEEDED --------------

### CODE REVIEW

File: src/MCP23017.cpp [hunk 1]
```diff
@@ -7,40 +7,79 @@ MCP23017::MCP23017(uint8_t address, TwoWire& bus) {
...
+	//IOCON.MIRROR = 	0 : use configureInterrupt 
+	//IOCON.SEQOP  = 	1 : sequential operation disabled, address pointer does not increment
+	//IOCON.DISSLW = 	0 : slew rate enabled
+	//IOCON.HAEN   = 	0 : hardware address pin is always enabled on 23017
+	//IOCON.ODR    = 	0 : open drain output
+	//IOCON.INTPOL = 	0 : interrupt active low
+	//IOCON bit 0 is spare
 	writeRegister(MCP23017_REGISTER::IOCON, 0b00100000);
 
 	//enable all pull up resistors (will be effective for input pins only)
```
Comment (blemasle) [inline src/MCP23017.cpp:22]:
```suggestion
	//enable pull up resistors (will be effective for input pins only)
```

File: src/MCP23017.cpp [hunk 2]
```diff
@@ -7,40 +7,79 @@ MCP23017::MCP23017(uint8_t address, TwoWire& bus) {
...
-	iodir = readRegister(iodirreg);
-	if(mode == OUTPUT) iodir &= ~_BV(pin);
-	else iodir |= _BV(pin);
+	uint8_t iodir = readRegister(iodirreg);
+	uint8_t pol   = readRegister(polreg);
+	uint8_t pull  = readRegister(pullreg);
+
+	if(mode == OUTPUT) 
+	{
+		iodir |= _BV(pin);
```
Comment (blemasle) [inline src/MCP23017.cpp:53]: Why did you invert the iodir assignation logic between `OUTPUT` and `INPUT` ? An `OUTPUT` pin must be cleared according to the datasheet, so `iodir &= ~_BV(pin);` was the correct way of handling this.

File: src/MCP23017.cpp [hunk 3]
```diff
@@ -7,40 +7,79 @@ MCP23017::MCP23017(uint8_t address, TwoWire& bus) {
...
+		pol   &= ~_BV(pin);
+		pull  &= ~_BV(pin);
+	} 
+	else
+	{
+		iodir &= ~_BV(pin);
```
Comment (blemasle) [inline src/MCP23017.cpp:59]: Same remark as above :)

**Review by blemasle — COMMENTED** [review]
It seems to me that there is a mistake in your code. Feel free to correct me If I'm wrong :)

File: src/MCP23017.cpp (hunk 2 above)
Comment (Beerlesklopfer) [reply src/MCP23017.cpp:53]: I did it first as you suggested, but the the result was just inverted as expected. This way solved that problem. My wiring was as described in the new example.

File: src/MCP23017.cpp (hunk 3 above)
Comment (Beerlesklopfer) [reply src/MCP23017.cpp:59]: me too 👍

File: src/MCP23017.cpp (hunk 2 above)
Comment (blemasle) [reply src/MCP23017.cpp:53]: I'm lost with that change. The code you're suggesting was initially the one I wrote, before it was corrected by #4.
I'll try to have a deeper look at this myself this weekend and sort it out once and for all. If that's true, then the datasheet is lying ^^

Comment (blemasle) [reply src/MCP23017.cpp:53]: So, I finally had the chance to get an mcp23017 out of the closet and test the library again.
I stick with my version of iodir, and this is very easy to prove :

I have a bunch of input switches on port B (active high), and some leds on port A. 

```cpp
#include <Wire.h>
#include <MCP23017.h>

#define MCP23017_ADDR 0x20
MCP23017 mcp = MCP23017(MCP23017_ADDR);

void setup() {
    Wire.begin();
    Serial.begin(115200);
    
    mcp.init();
    mcp.portMode(MCP23017_PORT::A, 0);         //Port A as output
    mcp.portMode(MCP23017_PORT::B, 0b11111111);//Port B as input

    mcp.writeRegister(MCP23017_REGISTER::GPIOA, 0x00); //Reset port A 
    mcp.writeRegister(MCP23017_REGISTER::GPIOB, 0x00); //Reset port B

    mcp.writeRegister(MCP23017_REGISTER::GPPUA, 0x00); //No pull ups on port A
    mcp.writeRegister(MCP23017_REGISTER::GPPUB, 0x00); //No pull ups on port B

    mcp.pinMode(7, INPUT);         //Pin 7 as input

    uint8_t conf = mcp.readRegister(MCP23017_REGISTER::IODIRA);
    Serial.print("IODIRA : ");
    Serial.print(conf, BIN);
    Serial.println();    
}


void loop() {
    uint8_t currentB;

    currentB = mcp.readPort(MCP23017_PORT::B);
    mcp.writePort(MCP23017_PORT::A, currentB);
}
```

> IODIRA : 10000000

In the example above, the led on pin 7 is not configured as an input, so it does not lit, whereas every other led do lit. This is with the current version of the library. Integrating your changes on `IODIR` gives

> IODIRA : 0

And all the leds are working (which proves that pin 7 is not configured as an input.

Comment (blemasle) [reply src/MCP23017.cpp:53]: Solved by v2.0.0. Feel free to reopen the discussion if you still disagree on the resolution. But in this is case, please read the comments on your example first. Your sketch has mistakes that clearly indicates you misused the library :)

Thanks for the suggestion, glad to see it finally made it to the library ! 👍

File: examples/InvertedInputs/InvertedInputs.ino [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+  Wire.begin();
+  mcp.begin(0x00, 0x00);
+  mcp.interruptMode(MCP23017_INTMODE::SEPARATED);
+  mcp.interrupt(MCP23017_PORT::A, CHANGE);
+  mcp.interrupt(MCP23017_PORT::B, CHANGE);
+
+  bool error = false;
+
+  // PORT A is input inverted with pullups
+  mcp.portMode( MCP23017_PORT::A, INPUT, 0xFF, 0xFF);
```
Comment (blemasle) [inline examples/InvertedInputs/InvertedInputs.ino:71]: The second parameter of port mode use an `uint8_t`, each bit represents the direction of the corresponding pin. As the MCP23017 use 1 for input, and 0 for output, here you're actually setting the port A to be an output port entirely.
```suggestion
  mcp.portMode( MCP23017_PORT::A, 0xFF, 0xFF, 0xFF); 
```
Would have produced the result you were looking for.

File: examples/InvertedInputs/InvertedInputs.ino [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+  mcp.interrupt(MCP23017_PORT::B, CHANGE);
+
+  bool error = false;
+
+  // PORT A is input inverted with pullups
+  mcp.portMode( MCP23017_PORT::A, INPUT, 0xFF, 0xFF); 
+  mcp.clearInterrupts();
+
+  // PORT B is output
+  mcp.portMode(MCP23017_PORT::B, OUTPUT);
```
Comment (blemasle) [inline examples/InvertedInputs/InvertedInputs.ino:75]: Same as above, `OUTPUT` on the arduino is `1`, to here you're only setting the first pin of the B port as an input.

```suggestion
  mcp.portMode(MCP23017_PORT::B, 0x00);
```

**Pull request closed without merging**